File: Daily/PY/Leetcode725-分隔链表.py
from typing import List


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def splitListToParts(self, head: ListNode, k: int) -> List[ListNode]:
        p,n = head,0
        while p:
            n += 1
            p = p.next
        p,ans = head,[]
        for i in range(k):
            base,mod = n // k,n % k
            if i + 1 <= mod:
                base += 1
            ans.append(p)
            for i in range(base - 1):
                p = p.next
            if p:
                q = p.next
                p.next = None
                p = q
        return ans

File: Leetcode/sbgo/Leetcode725-分隔链表.go
package sbgo

func splitListToParts(head *ListNode, k int) []*ListNode {
	n, p := 0, head
	for p != nil {
		n++
		p = p.Next
	}
	p = head
	ans := []*ListNode{}
	for i := 0; i < k; i++ {
		base, mod := n/k, n%k
		if i+1 <= mod {
			base++
		}
		ans = append(ans, p)
		for j := 0; j < base-1; j++ {
			p = p.Next
		}
		if p != nil {
			q := p.Next
			p.Next = nil
			p = q
		}
	}
	return ans
}
